src/gpuutils.py

```diff
@@ -0,0 +1,87 @@
+"""Utilities for measuring GPU performance."""
+
+from typing import Optional
+
+from logging import getLogger
+
+from pynvml import *
+import psutil
+
+import pandas as pd
+
+import time
+
+
+logger = getLogger(__file__)
+
+
+def gather_cpu_gpu_metrics(
+    interval: int | pd.Timedelta = 5,
+    iterations: int = 12,
+    *,
+    gather_for_time: Optional[pd.Timedelta] = None,
+    log_interval: Optional[int] = None,
```

Comment:
Never seen this either: what does specifying this 'Optional' type offer over just providing a default, which implies that it is optional?

Reply:
`Optional[type]` is equivalent to `type | None`, so it's specifically for cases where you either pass a value of type `type` or the default is `None`.